functorch/_src/aot_autograd.py

```diff
@@ -567,7 +1105,106 @@ def backward(ctx, *flat_args):
+
+        outs = CompiledFunction.apply(*no_dupe_args_with_synthetic_bases)
+
+        # Step 3: After running the compiled fw, apply updates to mutated inputs
+        if CompiledFunction.num_mutated_inputs > 0:
+            assert len(outs) == CompiledFunction.num_mutated_data_inputs + CompiledFunction.num_non_aliased_outs
+            # Calling convention: mutated inputs show up first.
+            updated_inputs = outs[:CompiledFunction.num_mutated_data_inputs]
+            fw_outs = outs[CompiledFunction.num_mutated_data_inputs:]
+            curr_mutated_inpt_idx = 0
+            curr_metadata_only_mutated_inpt_idx = 0
+            for inpt_idx, mutation_type in enumerate(CompiledFunction.fw_metadata.mutated_input_info):
+                if mutation_type == MutationType.none:
+                    continue
+                original_inpt = no_dupe_args[inpt_idx]
+                if mutation_type == MutationType.metadata_only:
+                    expected_metadata = CompiledFunction.fw_metadata\
+                        .metadata_mutation_input_info[curr_metadata_only_mutated_inpt_idx]
+                    curr_metadata_only_mutated_inpt_idx += 1
+                    original_inpt.as_strided_(
+                        expected_metadata.size(), expected_metadata.stride(), expected_metadata.storage_offset())
+                else:
+                    updated_inpt = updated_inputs[curr_mutated_inpt_idx]
+                    curr_mutated_inpt_idx += 1
+                    # TODO: handle resize_() on inputs to a larger size.
+                    # This is actually non-trivial to detect, so we should probably just handle it
+                    # (or make dynamo detect).
+                    # We can't just check of original_inpt.storage_size != updated_inpt.storage_size,
+                    # Because the original_inpt might be a view of some larger tensor,
+                    # and updated_inpt is always densely packed.
+                    if original_inpt.size() != updated_inpt.size() \
+                            or original_inpt.stride() != updated_inpt.stride() \
+                            or original_inpt.storage_offset() != updated_inpt.storage_offset():
+                        # Functionalization can't easily tell us if an input had BOTH its metadata actual data mutated.
+                        # So we check if metadata needs to be mutated here manually.
+                        original_inpt.as_strided_(updated_inpt.size(), updated_inpt.stride(), updated_inpt.storage_offset())
+                    original_inpt.copy_(updated_inpt)
+        else:
+            fw_outs = outs
+
+        # Step 4: Manually regenerate any outputs that are aliased to inputs, instead of
+        # compiling them.
+        if CompiledFunction.num_aliased_outputs > 0:
+            assert CompiledFunction.num_aliased_outputs + len(fw_outs) == len(CompiledFunction.fw_metadata.aliased_output_info)
+            fw_outs_including_aliases = []
```

Comment:
it might be profitable to avoid allocating a new list, and instead modify the old list in place. idk, maybe it doesn't matter